1_dynamic/future.k

```diff
@@ -16,4 +18,465 @@ endmodule
+*/
+
+  syntax ObjectVal ::= object(Int,Id,Id)
+  syntax Val ::= ObjectVal
+
+/* 
+  Redefinition of the semantics for new.
+  - an instance of the class is created (<object> cell) and
+    an unique identifier is assigned to it
+  - then the "$clinit" special method is called
+    as an ususal method to do class initialization
+  - once that is done, the constructor is called
+  - Finally, the result is the identifier of the new object
+  - Both class initialization and constructor are blocking to ensure
+    the object reference is not exposed during initialization.
+*/
+  syntax Id ::= "$clinit" 
+
+  rule <k> new Class:Id(Vs:Vals)
+           => object(Id,Class,Class) . $clinit(.Exps); ~> object(Id,Class,Class).Class(Vs); ~> object(Id,Class,Class) ...</k>
+       <nextObjectId> Id:Int => Id +Int 1 </nextObjectId>
+       (. => <object>...
+                <location> Id </location>
+                <baseClass> Class </baseClass>
+                <active-task>... <futureId> -1 </futureId> ...</active-task>
+             ...</object>)
+
+
+/*
+  The redefinition of the semantics for this:
+  - this returns a reference to the current object.
+*/
+
+  rule <k> this => object(Id,BClass,Class) ...</k>
+       <location> Id:Int </location>
+       <crntClass> Class:Id </crntClass>
+       <baseClass> BClass:Id </baseClass>
+
+/*
+  The semantics of a future call.  A future call creates a new future
+  and "returns" the identifier of the new future.
+  Elements needed:
+  * the object identifier (where the method is to be called)
+  * the class from which the method should be called
+    - currently used mainly for super
+    - could also be useful to implement other method lookup mechanisms
+  * the name of the method to be called
+  * the (already evaluated) arguments to be passed to the method.
+*/
+
+  // the arguments are strict such that they are evaluated first
+  context _:ObjectVal ! Method:Id(HOLE)
+
+  rule <k> object(OId,Class,_) ! Method:Id(Vs:Vals)
+        => future(F) ...</k>
+       (. => <future>...
+               <fid> F </fid> <oid> OId </oid>
+               <closure> lookupMethod(Class,Method) </closure>
+               <arguments> Vs </arguments>
+             ...</future>)
+       <nextFutureId> F:Int => F +Int 1 </nextFutureId>
+
+  // future identifiers
+  syntax Val ::= future(Int)
+  syntax KItem ::= lookupMethod(Id,Id)
+
+/* Looking up the location of a future method by name
+  * The method is looked up in the envStack of the object,
+    starting from specified class
+  * can be done in parallel with any existing tasks
+*/
+  rule <future>
+         <fid> F:Int </fid>
+         <oid> Id:Int </oid>
+          <closure>
+             lookupMethod(Class,Method)
+          => lookupThis(<envStack> ES </envStack>, Class, Method)
+          </closure>
+         <arguments> Vs:Vals </arguments>
+         <state> sleeping </state>
+       </future>
+       <location> Id </location>
+       <envStack> ES:List </envStack>
+
```

Comment:
Cele doua reguli de mai sus nu pot fi grupate in una singura?
A doua regula pare sa fie doar evaluarea unei functii, ce se presupune a fi o actiune spontana.